consensus/hotstuff/timeoutcollector/aggregation.go

```diff
@@ -0,0 +1,167 @@
+package timeoutcollector
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// signerInfo holds information about a signer, its public key and weight
+type signerInfo struct {
+	pk     crypto.PublicKey
+	weight uint64
+}
+
+// WeightedMultiMessageSignatureAggregator implements consensus/hotstuff.WeightedMultiMessageSignatureAggregator.
+// It aggregates BLS signatures for many messages from different signers.
+// Only public keys needs to be agreed upon upfront.
+// Each signer is allowed to sign at most once.
+// Aggregation uses BLS scheme. Mitigation against rogue attacks is done using Proof Of Possession (PoP)
+// This module does not verify PoPs of input public keys, it assumes verification was done outside this module.
+// Implementation is thread-safe.
+type WeightedMultiMessageSignatureAggregator struct {
```

Comment:
I really commend your effort to implement a reusable component. But as [Tarak has pointed out](https://github.com/onflow/flow-go/pull/2243#discussion_r843390102), this is quite non-trivial in general. For such a component to be universally useful _and safe_, it would need to
* > output should have the information that can be shared with another party and allow that party to verify the aggregated signature.
* It should guarantee that any aggregate it returns is correct (defensive programming). 

If achieving this is possible with reasonable overhead, I am happy. Though, I feel that we have designed a component with a general API, which is easily to misuse, requires explicit _extra_ actions to guarantee result correctness (check of the aggregate must be explicitly added by the engineer), and does not retain the necessary information to validate its result. 

### Suggestion
In addition, note that timeout aggregation is actually a _mixture_, where _all_ replicas sign the _same_ current view but different `QC.Views`. This is important, because we explicitly utilize the knowledge that the current view is the same for all signs for space efficiency. 
In other words, this signature scheme is super specific for timeout aggregation. Therefore, I think it warrants its own special-purpose aggregator. 

I feel we should rename this struct to `TimeoutSignatureAggregator` and include all the timeout-specific details:
* it knows the view
* When we add a timeout, we feed in: `SignerID`, `SigData`, and `HighestQC.View`.
* We add timeouts via `VerifyAndAdd`

The `TimeoutSignatureAggregator` would then:
* guarantee that it always returns a valid aggregate
* along with the aggregate sig return the `TOHighQCViews` slice (which we can directly use in the TC)
* own all the information that is required to verify the produced signature

Reply:
I think the suggestion here corresponds to the option 2 in my suggestion [here](https://github.com/onflow/flow-go/pull/2243#discussion_r860016309).